Implement LFU as default cache along with memory pool

qwe661234 opened the pull request:
We have implemented an adaptive replacement cache to reduce memory usage. However, the complexity of the replacement policy of ARC can lead to a decrease in overall performance. As a result, we implemented the LFU cache, which performed better, but we still retained the ARC for future use. Additionally, we imported the memory pool we developed to limit the memory usage of both caches.

The statistics below illustrate the performance of the LFU cache and ARC while running the CoreMark benchmark, indicating that the LFU cache outperforms ARC.

|Test| CoreMark(Iterations/Sec)  |
|--------| -------- | 
| ARC | 1123.776 |
| LFU Cache  |  1155.174  |

jserv commented on src/cache.c:
```diff
@@ -10,14 +10,17 @@
 static uint32_t cache_size, cache_size_bits;
-
+struct mpool *cache_mp;
+#define ARC
```
Check this definition.

jserv commented on src/cache.c:
```diff
@@ -10,14 +10,17 @@
 
 static uint32_t cache_size, cache_size_bits;
-
+struct mpool *cache_mp;
```
Should be declared as `static`.

jserv commented on src/feature.h:
```diff
@@ -48,6 +48,11 @@
 
+/* Import adaptive replacement cache to manage block */
+#ifndef RV32_FEATURE_ARC
+#define RV32_FEATURE_ARC 1
```
Make it disabled by default.

jserv commented on mk/tests.mk:
```diff
@@ -9,15 +9,22 @@ CACHE_OBJS := $(addprefix $(CACHE_BUILD_DIR)/, $(CACHE_OBJS))
-
+ifeq ($(ENABLE_ARC), 1)
+CACHE_CHECK_ELF_FILES := \
+	ARC/cache-new \
```
Lowercase should be used for the directory names. i.e., `arc/new` and `arc/put`.

jserv commented on Makefile:
```diff
@@ -40,6 +40,10 @@ ifeq ($(call has, EXT_F), 1)
 endif
 
+# Enable adaptive replacement cache policy, default is LRU 
+ENABLE_ARC ?= 0
```
Enable `ENABLE_ARC=1` in GitHub Actions for comprehensive builds.

jserv commented on src/cache.c:
```diff
@@ -393,21 +474,63 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
                                   arc_entry_t)
+#endif
+#else
+    for (int i = 0; i < 1000; i++) {
```
Why `1000`?

qwe661234 replied:
This would modify with threshold we set, maybe I define a marco as threshold.

jserv commented on mk/tests.mk:
```diff
@@ -9,15 +9,22 @@ CACHE_OBJS := $(addprefix $(CACHE_BUILD_DIR)/, $(CACHE_OBJS))
 deps += $(CACHE_OBJS:%.o=%.o.d)
 
-
+ifeq ($(ENABLE_ARC), 1)
```
Add comments about the relationship between LFU and ARC.

jserv commented on src/cache.c:
```diff
@@ -61,17 +66,31 @@ typedef struct {
     uint32_t list_size[N_CACHE_LIST_TYPES];
+    uint32_t lru_capacity;
+#else
+    struct list_head *lists[1000];
```
Avoid hardcoded `1000`.

jserv commented on src/cache.c:
```diff
@@ -10,14 +10,18 @@
 #define GOLDEN_RATIO_32 0x61C88647
 #define HASH(val) \
     (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) & (cache_size - 1)
+#define THRESHOLD 1000
```
Add comments.

jserv commented on src/cache.c:
```diff
@@ -80,6 +100,11 @@
     head->prev = head;
 }
 
+static inline int list_empty(const struct list_head *head)
```
Wrap with macro `!RV32_HAS(ARC)`.

jserv commented on Makefile:
```diff
@@ -40,6 +40,10 @@ ifeq ($(call has, EXT_F), 1)
 endif
 
+# Enable adaptive replacement cache policy, default is LRU 
+ENABLE_ARC ?= 1
```
By default, it is disabled.

jserv reviewed (changes requested):
Add corresponding build tests in GitHub Actions.

jserv commented on .github/workflows/main.yml:
```diff
@@ -18,6 +18,8 @@ jobs:
             make check
             make tests
             make misalign
+            make clean ENABLE_ARC=1
```
Should be `make ENABLE_ARC=1 clean tests`

qwe661234 replied:
`make ENABLE_ARC=1 clean all` for compile ARC version `cache.o`, `make ENABLE_ARC=1` tests for ARC tests.

jserv commented on Makefile:
```diff
@@ -55,7 +59,7 @@ $(OUT)/syscall_sdl.o: CFLAGS += $(shell sdl2-config --cflags)
 endif
 
-ENABLE_GDBSTUB ?= 1
+ENABLE_GDBSTUB ?= 0
```
Don't change this.

jserv commented on src/cache.c:
```diff
@@ -30,14 +36,15 @@ static uint32_t cache_size, cache_size_bits;
  * we increase T1's size while decreasing T2. But, if the cache hit occurs in
  * B2, we would increase the size of T2 and decrease the size of T1.
  */
+
```
Minimize the necessary changes. Don't add a blank line here.

jserv commented on src/cache.c:
```diff
@@ -323,13 +387,35 @@ void *cache_get(cache_t *cache, uint32_t key)
+    if (!entry || entry->key != key)
+        return NULL;
+
+    /* We would translate the block with a frequency of more thanTHRESHOLD */
```
Improve the writing.

jserv commented on src/cache.c:
```diff
@@ -393,21 +479,63 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
     hlist_add_head(&new_entry->ht_list, &cache->map->ht_list_head[HASH(key)]);
 
     CACHE_ASSERT(cache);
+#else
```
Add comment `!RV32_HAS(ARC)`.

jserv commented on src/cache.c:
```diff
@@ -393,21 +479,63 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
     CACHE_ASSERT(cache);
+#else
+    assert(cache->list_size <= cache->capacity);
+    /* Before adding new element to cach, we should check the status
```
Improve the writing. Check with ChatGPT.

jserv commented on src/cache.c:
```diff
@@ -393,21 +479,63 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
         list_for_each_entry_safe (entry, safe, cache->lists[i], list,
                                   arc_entry_t)
+#endif
+#else
```
Ditto. Add comment `!RV32_HAS(ARC)`.

jserv commented on src/cache.c:
```diff
@@ -224,12 +256,41 @@ cache_t *cache_create(int size_bits)
+    cache->lru_capacity = cache_size / 2;
+    cache_mp =
+        mpool_create(cache_size * 2 * sizeof(arc_entry_t), sizeof(arc_entry_t));
+#else
```
Should be `#else /* !RV32_HAS(ARC) */`

jserv commented on src/cache.c:
```diff
@@ -323,13 +386,35 @@ void *cache_get(cache_t *cache, uint32_t key)
     }
 
     CACHE_ASSERT(cache);
+#else
```
Should be `#else /* !RV32_HAS(ARC) */`

jserv commented on src/cache.c:
```diff
@@ -10,14 +10,20 @@
 #define GOLDEN_RATIO_32 0x61C88647
 #define HASH(val) \
     (((val) * (GOLDEN_RATIO_32)) >> (32 - (cache_size_bits))) & (cache_size - 1)
+/* THRESHOLD is set to identify hot spots. Once the frequency of use for a block
```
Be aware of consistent style. i.e.,
```c
/* THRESHOLD is set to identify hot spots. Once the frequency of use for a block
 * exceeds the THRESHOLD, it will be translated into machine code.*
 * /
```

The statement "translated into machine code" is not accurate.

jserv replied:
Be aware of consistent style. i.e.,
```c
/* THRESHOLD is set to identify hot spots. Once the frequency of use for a block
 * exceeds the THRESHOLD, it will be translated into machine code.*
 * /
```

The statement "translated into machine code" is not accurate.

jserv commented on src/cache.c:
```diff
@@ -393,21 +480,62 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
     CACHE_ASSERT(cache);
+#else /* !RV32_HAS(ARC) */
+    assert(cache->list_size <= cache->capacity);
+    /* It is necessary to check the cache is full or not before adding a new
```
Make the comment style more consistent.

jserv commented on src/cache.c:
```diff
@@ -393,21 +480,62 @@ void *cache_put(cache_t *cache, uint32_t key, void *value)
+     * element */
+    if (cache->list_size == cache->capacity) {
+        for (int i = 0; i < THRESHOLD; i++) {
+            if (!list_empty(cache->lists[i])) {
```
Rewrite as following:
```c
        for (int i = 0; i < THRESHOLD; i++) {
            if (list_empty(cache->lists[i]))
                continue;
```

qwe661234 committed:
Refactor ARC into LFU and integrate mpool

We have implemented an adaptive replacement cache to reduce memory usage.
However, the complexity of the replacement policy of ARC can lead to a
decrease in overall performance. As a result, we implemented the LFU cache,
which performed better, but we still retained the ARC for future use.
Additionally, we imported the memory pool we developed to limit the memory
usage of both caches.

The statistics below illustrate the performance of the LFU cache and ARC
while running the CoreMark benchmark, indicating that the LFU cache outperforms ARC.

| Test |  CoreMark(Iterations/Sec)  |
|------+----------------------------|
| ARC  |         1123.776           |
| LFU  |         1155.174           |

jserv commented:
Thank @qwe661234 for contributing. Let's move forward to JIT compilation.
